pandas/core/dtypes/missing.py

```diff
@@ -54,6 +53,19 @@
 )
 from pandas.core.dtypes.inference import is_list_like
 
+if TYPE_CHECKING:
+    from pandas._typing import (
+        ArrayLike,
+        DtypeObj,
+        NDFrame,
+        NDFrameT,
+        Scalar,
+        npt,
+    )
```

Comment:
Thanks @twoertwein for the PR.

I didn't want to comment on this PR before to avoid "too many cooks" and in general if mypy is happy, so am I. (In some respects, because we have mypy we don't need to review certain aspects of typing as if it is not 100% correct, it will create issues down the line. I see it like a jigsaw puzzle that won't be complete until the last piece is in place, i.e. the codebase is 100% typed.)

However, I think I've seen that your preference is to import inside the TYPE_CHECKING also elsewhere? If this is correct can you explain the reasoning so that it helps when reviewing other contributors PRs.

AFAIK, we ensure all imports in pandas._typing are guarded so that they can be imported at the top level and for instance npt was added so that we could import without needing to add a TYPE_CHECKING block everywhere it was needed, otherwise we would just import from numpy directly and not include in pandas._typing?

Reply:
Good point! I imported it within the TYPE_CHECKING block because it is only needed for type checking . I will import it outside the TYPE_CHECKING block in next PRs.

Probably one reason why I prefer to put imports in this block is that there are some import cycles that prevent even mypy to function correctly: I think having `from pandas import Index` caused issues whereas `from pandas.core.indexes.base import Index` worked without issues (both inside the TYPE_CHECKING block).

Reply:
Yes, for import cycles it is definitely needed. My preference is also adding imports inside the type checking if added specifically in a typing PR for type annotations but this is difficult to enforce since say a refactor could remove the need for a top level import and it would not be removed because it is still used in type annotations then in theory the import should be moved to the type checking for consistency.

Reply:
> I think having `from pandas import Index` caused issues whereas `from pandas.core.indexes.base import Index` worked without issues (both inside the TYPE_CHECKING block).

my preference here is to use `from pandas import Index` when inside the type checking block. less typing! (the fingers on the keyboard type of typing not the type annotations type of typing!)

Reply:
I just tried replacing it with `from pandas import Index` and it worked, must have been something different that caused weird mypy issues.